rule R2_mapping:
    input:
        ref_genome=config["refs"][config["ref_genome"]]["bwa_mem2_index"],
        R2 = rules.multiSplit.output.split_R2
    output:
        R2sortbam = "processed/{sample}/{split}/{sample}.R2.sort.bam"
    threads:  config["resources"]["bwa_cpu_threads"],
    conda:"../envs/main_env.yaml",
    shell:"""

        bwa-mem2 mem -5SP -t {threads} {input.ref_genome} {input.R2} | samtools sort -@{threads} -o {output.R2sortbam} -

    """

rule bam2bed:
    input:
        R2sortbam = rules.R2_mapping.output.R2sortbam,
        hickit = config["softwares"]["HICKIT"],
        snp_file = config["refs"][config["ref_genome"]]["snp"],
    output:
        R2_bed = "processed/{sample}/{split}/{sample}.{split}.R2_5.bed.gz"
    conda:"../envs/main_env.yaml",
    shell:"""

        samtools view {input.R2sortbam} | xargs -I {{}} printf '{{}}\n%.0s' {{1..2}} | \
        {input.hickit}/hickit.js sam2seg -d 0 -v {input.snp_file} - | \
        cut -f 2 | awk '{{FS="!";OFS="\t"}}{{print $1,$2,$3,$5,$6,$4}}' | gzip > {output.R2_bed}

    """

rule generate_dedup_fragments:
    input:
        R2_bed = rules.bam2bed.output.R2_bed,
    output:
        fragments = "processed/{split}_all/{sample}.{split}.frag.bed.gz",
    log: "processed/{split}_all/{sample}.{split}.frag.log",
    #conda:"../envs/main_env.yaml",
    shell:"""
        set +u
        source ~/miniconda3/etc/profile.d/conda.sh
        conda activate hic2
        set -u
        
        python ./CHARM/CHARM_scripts/dedup_bed.py -i {input.R2_bed} -o {output.fragments} -t normal > {log} > {log}

        set +u
        conda deactivate
        set -u
        
    """

rule foramt_fragmenets:
    input:
        fragments = rules.generate_dedup_fragments.output.fragments,
    output:
        format_fragments = temp("processed/{split}_all/{sample}.{split}.format_frag.bed.gz")
    conda:"../envs/main_env.yaml",
    shell:"""
        zcat {input.fragments} | cut -f 1-3 | awk '{{OFS="\t"}}{{print $0"\t{wildcards.sample}\t1"}}' | gzip > {output.format_fragments}
    """

rule combine_fragments:
    input:
        fragments = expand("processed/{{split}}_all/{sample}.{{split}}.format_frag.bed.gz", sample=SAMPLES),
    output:
        fragments = "result/fragments/{split}.fragments.bgz",
    conda:"../envs/main_env.yaml",
    threads: 10
    shell:"""
        zcat {input.fragments} | sort -k1,1 -k2,2n | uniq |  bgzip -@ 5 > {output.fragments}
    """